src/Access/SettingsConstraints.cpp

```diff
@@ -154,95 +120,111 @@ void SettingsConstraints::merge(const SettingsConstraints & other)
         check(current_settings, change);
 }
 
+void SettingsConstraints::check(const Settings & current_settings, SettingsChanges & changes) const
+{
+    boost::range::remove_erase_if(
+        changes,
+        [&](SettingChange & change) -> bool
+        {
+            return !checkImpl(current_settings, const_cast<SettingChange &>(change), THROW_ON_VIOLATION);
+        });
+}
 
-void SettingsConstraints::clamp(const Settings & current_settings, SettingChange & change) const
+void SettingsConstraints::clamp(const Settings & current_settings, SettingsChanges & changes) const
 {
-    const String & name = change.name;
-    size_t setting_index = Settings::findIndex(name);
-    if (setting_index == Settings::npos)
-        return;
+    boost::range::remove_erase_if(
+        changes,
+        [&](SettingChange & change) -> bool
+        {
+            return !checkImpl(current_settings, change, CLAMP_ON_VIOLATION);
+        });
+}
 
-    Field new_value = Settings::valueToCorrespondingType(setting_index, change.value);
-    Field current_value = current_settings.get(setting_index);
 
-    /// Setting isn't checked if value wasn't changed.
-    if (current_value == new_value)
-        return;
+bool SettingsConstraints::checkImpl(const Settings & current_settings, SettingChange & change, ReactionOnViolation reaction) const
+{
+    const String & setting_name = change.name;
+
+    bool cannot_cast;
+    auto cast_value = [&](const Field & x) -> Field
+    {
+        cannot_cast = false;
+        if (reaction == THROW_ON_VIOLATION)
+            return Settings::castValueUtil(setting_name, x);
+        else
+        {
+            try
+            {
+                return Settings::castValueUtil(setting_name, x);
+            }
+            catch (...)
+            {
+                cannot_cast = true;
+                return {};
+            }
+        }
+    };
 
-    if (!current_settings.allow_ddl && name == "allow_ddl")
+    bool cannot_compare = false;
+    auto less = [&](const Field & left, const Field & right)
     {
-        change.value = current_value;
-        return;
+        cannot_compare = false;
+        if (reaction == THROW_ON_VIOLATION)
+            return applyVisitor(FieldVisitorAccurateLess{}, left, right);
+        else
+        {
+            try
+            {
+                return applyVisitor(FieldVisitorAccurateLess{}, left, right);
+            }
+            catch (...)
+            {
+                cannot_compare = true;
+                return false;
+            }
+        }
+    };
+
+    if (reaction == THROW_ON_VIOLATION)
+        manager->checkSettingNameIsAllowed(setting_name);
+    else if (!manager->isSettingNameAllowed(setting_name))
+        return false;
+
+    Field current_value, new_value;
+    if (current_settings.tryGet(setting_name, current_value))
+    {
+        /// Setting isn't checked if value has not changed.
+        if (change.value == current_value)
```

Comment:
This changes the behavior a little, now if you change the value of the setting to it's default it will not be reflected.

```
$ clickhouse-client --min_compress_block_size=65536 -q "select changed from system.settings where name = 'min_compress_block_size'"
0
```

*(And obviously also it won't be reflected in the query_log)*

Although via `clickhouse-local` it is `changed` (since it does not goes through the following code path):

```
$ clickhouse-local --min_compress_block_size=65536 -q "select changed from system.settings where name = 'min_compress_block_size'"
1
```

I guess it will be fine if it will be consistent, but it is different for `clickhouse-local` and `clickhouse-client`, and this is not fine, @vitlibar ?